### PR TITLE
Retry call caching centaur tests BT-271

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

**Review by mcovarr — APPROVED** [review]

File: centaur/src/main/scala/centaur/CromwellDatabase.scala [hunk 1]
```diff
@@ -10,10 +10,13 @@ trait CromwellDatabase {
 }
 
 object CromwellDatabase {
+
+  lazy val instance: CromwellDatabase = fromConfig(CentaurConfig.conf)
```
Comment (breilly2) [inline centaur/src/main/scala/centaur/CromwellDatabase.scala:14]: I'm not sure what effect it might have to change this object's lifecycle. Maybe worth a quick chat.

Comment (kshakir) [reply centaur/src/main/scala/centaur/CromwellDatabase.scala:14]: Discussed IRL. The `object ErrorReporters` already effectively had its own singleton of the database object. So no change is expected around `ErrorReporters`, which was not closing the database `instance`.

There is a line in this PR that also used the same `instance` for tests extending from `AbstractCromwellEngineOrBackendUpgradeTestCaseSpec`. `
From the discussion, there's no reason that each `AbstractCromwellEngineOrBackendUpgradeTestCaseSpec` needs its own connection pool(s... ackchyually one for engine, one for metadata). AbstractCromwellEngineOrBackendUpgradeTestCaseSpec` currently does not close its instance. The `AbstractCromwellEngineOrBackendUpgradeTestCaseSpec` test \_do\_ drop and re-create the database \_schemas\_. However, the tests can do that existing work through connections retrieved from the singleton connection pool `instance` versus creating their own connection pool(s).

In CI testing all the tests meet previous regression expectations. If we decide to not use a singleton connection pool(s) in the future for any reason we can switch `AbstractCromwellEngineOrBackendUpgradeTestCaseSpec` back to creating its own separate connections.

File: centaur/src/main/scala/centaur/test/standard/CentaurTestCase.scala [hunk 2]
```diff
@@ -52,14 +52,6 @@ case class CentaurTestCase(workflow: Workflow,
   }
 
   def containsTag(tag: String): Boolean = testOptions.tags.contains(tag)
-
-  def reliesOnCallCachingMetadataVerification: Boolean = {
```
Comment (breilly2) [inline centaur/src/main/scala/centaur/test/standard/CentaurTestCase.scala:56]: I don't know what the implication of removing this is. Is this related to references like this?

http://github.com/broadinstitute/cromwell/blob/187c49defd1ef3c2df4133e6e9b9c92edfce79b4/centaur/src/main/resources/standardTestCases/bucket_name_with_no_trailing_slash.test#L11-L11

File: centaur/src/it/scala/centaur/AbstractCentaurTestCaseSpec.scala [hunk 3]
```diff
@@ -178,6 +175,10 @@ abstract class AbstractCentaurTestCaseSpec(cromwellBackends: List[String], cromw
       for {
         _ <- ErrorReporters.logFailure(testEnvironment, centaurTestException)
         r <- if (attempt < retries) {
+          centaurTestException
+            .workflowIdOption
+            .map(CromwellDatabaseCallCaching.clearCachedResults)
+            .getOrElse(IO.unit) *>
```
[the remark this answers is not in the document]
Comment (kshakir) [reply centaur/src/it/scala/centaur/AbstractCentaurTestCaseSpec.scala:181]: Oops. I knew I forgot to discuss something today, but I think we also discussed a lot of other stuff.

Anyways, here are some async links that discuss this operator that is an alias for `productR`:
- [Official terse docs](https://typelevel.org/cats-effect/api/2.x/cats/effect/IO.html#*%3E[B](another:cats.effect.IO[B]):cats.effect.IO[B]) on the cats effect 2.x implementation of operator
- [Blog post on the operator](https://blog.softwaremill.com/9-tips-about-using-cats-in-scala-you-might-want-to-know-e1bafd365f88#8d9b) in general
- If/when we upgrade to the latest and [greatest](https://www.youtube.com/watch?v=KZtVBtOrP50) 3.x one could use the more verbose [`productR`](https://typelevel.org/cats-effect/api/3.x/cats/effect/IO.html#productR[B](that:cats.effect.IO[B]):cats.effect.IO[B]) instead, but `productR` is \*not\* aliased in [2.x](https://typelevel.org/cats-effect/api/2.x/cats/effect/IO.html)

File: database/sql/src/main/scala/cromwell/database/slick/tables/CallCachingEntryComponent.scala [hunk 4]
```diff
@@ -50,18 +50,25 @@ trait CallCachingEntryComponent {
...
+      callCachingEntry <- callCachingEntries
+      if callCachingEntry.workflowExecutionUuid === workflowId
+      if callCachingEntry.callFullyQualifiedName === callFqn
+      if callCachingEntry.jobIndex === jobIndex
+    } yield callCachingEntry
   )
 
-  def callCachingEntryIdsForWorkflowId(workflowId: String) = {
-    for {
+  val callCachingEntryIdsForWorkflowId = Compiled(
```
[the remark this answers is not in the document]
Comment (kshakir) [reply database/sql/src/main/scala/cromwell/database/slick/tables/CallCachingEntryComponent.scala:61]: Functional.

You generally have two options for converting slick DSL into SQL:
- Every time you enter the slick code you convert it to parameterized SQL
- Only the first time you "[compile](https://scala-slick.org/doc/3.3.2/queries.html#compiled-queries)" the slick code into parameterized SQL and then keep the compiled result in memory

For some reason, this was doing the former. I suspect the implementer and reviewer just didn't think to cache the DSL-to-SQL conversion.

File: centaur/src/main/scala/centaur/test/standard/CentaurTestCase.scala (hunk 2 above)
Comment (kshakir) [reply centaur/src/main/scala/centaur/test/standard/CentaurTestCase.scala:56]: Yes, this PR is reverting https://github.com/broadinstitute/cromwell/pull/5322. That previous PR quickly gave up on tests like the one you linked. Test like `bucket_name_with_no_trailing_slash` are verifying metadata specific to call-caching results.

The net result of this PR: restore bq-report-and-retrying for any test that verifies call-caching results-- after invalidating the failed call-caching entries from the Cromwell database.

**Review by breilly2 — APPROVED** [review]
👍   Lovely!
